updater/sponsorships.py

```diff
@@ -50,82 +50,131 @@ def get_events(app):
     return sponsoreds, tb
 
 
+def applyV5(app, context_id):
+    # check the context id is linked
+    context = contexts[app['context']]
+    collection = context['collection']
+    c = db[collection].find({'contextId': context_id})
+    if c.empty():
+        db['sponsorships'].insert({
+            '_from': 'users/0',
+            '_to': 'apps/' + app['_key'],
+            'expireDate': int(time.time()) + 3600,
+            'contextId': context_id,
+        })
+        print(f"{context_id} is not linked to a brightid yet, a temporary sponsorship is added to be applied after user links contextId")
+        return
+
+    user = c.next()['user']
+    c = sponsorships.find({'_from': f'users/{user}'})
+    if not c.empty():
+        print("the user is sponsored before")
+        return
+
+    op = {
+        'name': 'Sponsor',
+        'app': app['_key'],
+        'id': user,
+        'timestamp': int(time.time() * 1000),
+        'v': 5
+    }
+
+    signing_key = ed25519.SigningKey(
+        base64.b64decode(app['sponsorPrivateKey']))
+    message = json.dumps(op, sort_keys=True,
+                         separators=(',', ':')).encode('ascii')
+    sig = signing_key.sign(message)
+    op['sig'] = base64.b64encode(sig).decode('ascii')
+    h = base64.b64encode(sha256(message).digest()).decode("ascii")
+    op['hash'] = h.replace('/', '_').replace('+', '-').replace('=', '')
+    op['state'] = 'init'
+    op['_key'] = op['hash']
+    operation = db['operations'].get(op['hash'])
+    if not operation:
+        db['operations'].insert(op)
+
+
+def applyV6(app, app_id):
+    c = sponsorships.find({'appId': app_id})
+    if c.empty():
+        db['sponsorships'].insert({
+            '_from': 'users/0',
+            '_to': 'apps/' + app['_key'],
+            'expireDate': int(time.time()) + 3600,
+            'appId': app_id,
+            'state': 'app',
+        })
+        return
+
+    sponsorship = c.next()
+    if sponsorship['state'] in ['app', 'done']:
+        print('the app id is sponsored before')
+        return
+
+    db['sponsorships'].update({
+        '_key': sponsorship['_key'],
+        'expireDate': None,
+        'state': 'done',
+        'timestamp': int(time.time() * 1000),
+    })
```

Comment:
Looking at this code, you'd need hidden knowledge of the value `client` as a possible value for the `state` element to make sense of it.

Instead of using one element with three states, using two element with two states each would make the system more transparent and self-explanatory.

For example, instead of using `state` (which doesn't say which states are possible, or even what they're for), you could have
`appHasAuthorized` <Boolean> 
and
`spendRequested` <Boolean>
This would make it obvious just by the names of the elements what they are for, and there are no hidden values, just `true` or `false`.

Reply:
It's done
https://github.com/BrightID/BrightID-Node/commit/ec0c8d48d4a04fdd765ca7abd834777162ea7902